File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {
        return dfs(root,root.val);
    }
    
    public int dfs(TreeNode node, int max){
        if (node is null ) return 0;
        //keep track of max at each level
        int biggest = max;

        int good = 0;
        if (node.val >= max){ //bigger or equal to prev node
            good++;
        }

        biggest = Math.Max(biggest, node.val);
        good += dfs(node.left,biggest);
        good += dfs(node.right,biggest);

        return good;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.res = []

        self.dfs(root, self.res)
        #1 indexed ( K - 1)
        return self.res[k - 1]

    def dfs(self, root, res):
        if not root:
            return 

        #InOrder = Left - Node - Right
        self.dfs(root.left, self.res)
        self.res.append(root.val)
        self.dfs(root.right,self.res)
        
    
    #perform inOrder traversal